cpp/include/cudf/utilities/error.hpp

```diff
@@ -77,37 +77,59 @@ struct fatal_cuda_error : public cuda_error {
  *
  * @code
- * CUDF_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch");
+ * // throws cudf::logic_error
+ * CUDF_EXPECTS(lhs.type() != rhs.type(), "Column type mismatch");
+ *
+ * // throws std::invalid_argument
+ * CUDF_EXPECTS(not cudf::is_nested(child_col.type()), std::invalid_argument,
+ *      "Nested types are not supported.");
  * @endcode
  *
- * @param[in] cond Expression that evaluates to true or false
- * @param[in] reason String literal description of the reason that cond is
- * expected to be true
- * @throw cudf::logic_error if the condition evaluates to false.
+ * @param[in] _condition Expression that evaluates to true or false
+ * @param[in] _expection_type The exception type to throw; must inherit
+ *     `std::exception`. If not specified (i.e. if only two macro
+ *     arguments are provided), defaults to `cudf::logic_error`
+ * @param[in] _what  String literal description of why the exception was
+ *     thrown, i.e. why `_condition` was expected to be true.
+ * @throw `_exception_type` if the condition evaluates to 0 (false).
  */
-#define CUDF_EXPECTS(cond, reason)                                  \
-  (!!(cond)) ? static_cast<void>(0)                                 \
-             : throw cudf::logic_error("cuDF failure at: " __FILE__ \
-                                       ":" CUDF_STRINGIFY(__LINE__) ": " reason)
+#define CUDF_EXPECTS(...)                                             \
+  GET_CUDF_EXPECTS_MACRO(__VA_ARGS__, CUDF_EXPECTS_3, CUDF_EXPECTS_2) \
+  (__VA_ARGS__)
+#define GET_CUDF_EXPECTS_MACRO(_1, _2, _3, NAME, ...) NAME
+#define CUDF_EXPECTS_3(_condition, _exception_type, _reason)               \
+  (!!(_condition)) ? static_cast<void>(0) : throw _exception_type          \
+  {                                                                        \
+    "cuDF failure at: " __FILE__ ":" CUDF_STRINGIFY(__LINE__) ": " _reason \
+  }
+#define CUDF_EXPECTS_2(_condition, _reason) CUDF_EXPECTS_3(_condition, cudf::logic_error, _reason)
 
 /**
  * @brief Indicates that an erroneous code path has been taken.
  *
- * In host code, throws a `cudf::logic_error`.
- *
- *
  * Example usage:
- * ```
+ * ```c++
+ * // Throws `cudf::logic_error`
  * CUDF_FAIL("Non-arithmetic operation is not supported");
- * ```
  *
- * @param[in] reason String literal description of the reason
+ * // Throws `std::invalid_argument`
+ * CUDF_FAIL("Unsupported type for operation", std::invalid_argument);
+ * ```
  */
-#define CUDF_FAIL(reason) \
-  throw cudf::logic_error("cuDF failure at: " __FILE__ ":" CUDF_STRINGIFY(__LINE__) ": " reason)
+#define CUDF_FAIL(...)                                       \
+  GET_CUDF_FAIL_MACRO(__VA_ARGS__, CUDF_FAIL_2, CUDF_FAIL_1) \
+  (__VA_ARGS__)
+#define GET_CUDF_FAIL_MACRO(_1, _2, NAME, ...) NAME
+#define CUDF_FAIL_2(_what, _exception_type)      \
+  /*NOLINTNEXTLINE(bugprone-macro-parentheses)*/ \
+  throw _exception_type{"cuDF failure at:" __FILE__ ":" CUDF_STRINGIFY(__LINE__) ": " _what};
```

Comment:
Maybe for DRY this should just call `CUDF_EXPECTS(false, _exception_type, _what)`?